Proton/Sources/Helpers/NSAttributedString+Range.swift

```diff
@@ -126,4 +126,14 @@ public extension NSAttributedString {
         guard location < length else { return nil }
         return attribute(attributeKey, at: location, effectiveRange: nil) as? T
     }
+    
+    /// Alternative to `attributedSubstring(from:_).string`
+    /// Avoids allocating `NSAttributedString` and all the attributes for that range, only to ignore the range.
+    func substring(from range: NSRange) -> String {
+        guard range.upperBound <= length else {
+            assertionFailure("Substring is out of bounds")
+            return ""
+        }
+        return (string as NSString).substring(with: range)
```

Comment:
Are you saying this part `(string as NSString)` is free because underlying storage is `NSString`? Does Apple have it documented somewhere or you made a test?

Reply:
My understanding here is that `NSAttributedString::string` returns an `NSString`. But the Swift runtime bridges it back to `Swift.String` automatically. And as you know, this `as` isn't really a "cast" so much a bridging it back to `NSString`. 

But the point of this code is that I need access to `.substring(with: NSRange)`. If I just use `.string` then I need a `Range<String.Index>` which is too hard (read "safer"). 

I could not figure out if this bridging to `String` and back to `NSString` is no-op'd, or otherwise "free".

The performance improvement in this PR has nothing to do with this cast. The performance is that we're not allocating and bridging the NSDictionaries for each range and all the keys and all of the attributes to Swift. I noticed it calls our `PRTextStorage` a lot for attributes(at:) in order to build these attributedSubstring. 


Alternatively, if we are concerned about the double bridging, perhaps we could write this in Objective-C?